### ISSUE TITLE
Parsing String values crashes ESP8266

### ISSUE THREAD

**Issue by JeffAshtonCT** [bblanchon/ArduinoJson#104]
Also noticing on the ESP8266, that parsing anything but a statically defined const char*, kills the parser.

```
#include <ArduinoJson.h>

void setup() {

  Serial.begin( 115200 );
  delay( 10 );
  Serial.println( "" );

  StaticJsonBuffer<200> jsonBuffer;
    
  String json = "{\"mqtt.host\":\"mqtt.test.com\",\"mqtt.port\":1883}";
    
  Serial.println( "Parsing json string" );
  JsonObject& root = jsonBuffer.parseObject( json );
  
  if( root.success() ) {
    Serial.println( "parsed" );
  } else {
    Serial.println( "failed" );
  }
}

void loop() {
}
```

Results in

```
Parsing json string

Exception (9):
epc1=0x402021cd epc2=0x00000000 epc3=0x00000000 excvaddr=0x3ffe9d67 depc=0x00000000

ctx: cont 
sp: 3ffe9b40 end: 3ffe9e30 offset: 01a0

>>>stack>>>
3ffe9ce0:  3fffdc20 00000013 3ffe8de4 402039a9  
3ffe9cf0:  3ffe8968 00000000 3ffe8de4 3ffe9e5c  
3ffe9d00:  3fffdc20 00000000 3ffe8de4 40202216  
3ffe9d10:  3ffe9d30 3ffe9d34 3ffe9d34 4020450a  
3ffe9d20:  3fffdc20 00000000 3ffe8de4 4020214e  
3ffe9d30:  3ffe8878 716d227b 682e7474 2274736f  
3ffe9d40:  716d223a 742e7474 2e747365 226d6f63  
3ffe9d50:  716d222c 702e7474 2274726f 3838313a  
3ffe9d60:  00007d33 00000000 00000000 00000000  
3ffe9d70:  00000000 00000000 00000000 00000000  
3ffe9d80:  00000000 00000000 00000000 00000000  
3ffe9d90:  00000000 00000000 00000000 00000000  
3ffe9da0:  00000000 00000000 00000000 00000000  
3ffe9db0:  00000000 00000000 00000000 00000000  
3ffe9dc0:  00000000 00000000 00000000 00000000  
3ffe9dd0:  00000000 00000000 00000000 00000000  
3ffe9de0:  00000000 00000000 00000000 00000000  
3ffe9df0:  00000000 00000000 00000000 00000037  
3ffe9e00:  3fff38a8 0000002e 0000002e 00000000  
3ffe9e10:  00000000 00000000 3ffe9e54 40201a07  
3ffe9e20:  00000000 00000000 3ffe8e10 40100378  
<<<stack<<<
```

**Comment by JeffAshtonCT** [comment]
The issue occurs here 
[JsonBuffer.cpp#L22](https://github.com/bblanchon/ArduinoJson/blob/master/src/JsonBuffer.cpp#L22) when it tries to create the new JsonObject.

**Comment by bblanchon** [comment]
I'm not able to reproduce this bug.

Please dig deeper in the `new` operator, in [JsonBufferAllocated.hpp, line 16](https://github.com/bblanchon/ArduinoJson/blob/master/include/ArduinoJson/Internals/JsonBufferAllocated.hpp#L16).

The `StaticJsonBuffer` is probably too small to hold a copy of the JSON source and the object tree, but that's not a reason to crash like this.

Please also try to increase the size of the `StaticJsonBuffer` to see if it improves.

Also, please note that using `String` instead of `char[]` will have a huge negative impact on program performance and code size. The JSON input will be duplicated (you saw the call to `strdup`) and you'll pay the cost of `String`'s internal allocation and duplication.

**Comment by bblanchon** [comment]
I fixed a bug which is triggered when the `StaticJsonBuffer` is too small to hold a copy of the JSON source.

It could be the same bug.

Please try [branch issue104](https://github.com/bblanchon/ArduinoJson/tree/issue104) and tell me if it works now.

**Comment by bblanchon** [comment]
I merge it into master.
Please give it a try.

**Comment by JeffAshtonCT** [comment]
Still faulting with the latest

```
Parsing json string

Exception (9):
epc1=0x4020226d epc2=0x00000000 epc3=0x00000000 excvaddr=0x3ffe9727 depc=0x00000000

ctx: cont 
sp: 3ffe94f0 end: 3ffe97f0 offset: 01a0

>>>stack>>>
3ffe9690:  3ffe9848 00000012 3ffe96d0 40202a25  
3ffe96a0:  3fffdc20 00000013 3ffe9848 402031f5  
3ffe96b0:  3ffe8518 00000000 3ffe9848 3ffe981c  
3ffe96c0:  3fffdc20 3fff0a08 3ffe9848 402022c4  
3ffe96d0:  3ffe96f0 3ffe96f4 3ffe96f4 4020570a  
3ffe96e0:  3fffdc20 3fff0a08 3ffe9848 402021ee  
3ffe96f0:  3ffe8478 716d227b 682e7474 2274736f  
3ffe9700:  716d223a 742e7474 2e747365 226d6f63  
3ffe9710:  716d222c 702e7474 2274726f 3838313a  
3ffe9720:  00007d33 3fffdc20 3ffe981c 00000030  
3ffe9730:  00000000 00000000 00000000 00000000  
3ffe9740:  00000000 00000000 00000000 00000000  
3ffe9750:  00000000 00000000 00000000 00000000  
3ffe9760:  00000000 00000000 00000000 00000000  
3ffe9770:  00000000 00000000 00000000 00000000  
3ffe9780:  00000000 00000000 00000000 00000000  
3ffe9790:  00000000 00000000 00000000 00000000  
3ffe97a0:  00000000 00000000 00000000 00000000  
3ffe97b0:  00000000 00000000 00000000 00000037  
3ffe97c0:  3fff0a08 0000002e 0000002e 00000000  
3ffe97d0:  00000000 00000000 3ffe9814 40201a43  
3ffe97e0:  00000000 00000000 3ffe87d0 40100374  
<<<stack<<<

 ets Jan  8 2013,rst cause:2, boot mode:(3,6)
```

**Comment by bblanchon** [comment]
I ordered [this](https://www.sparkfun.com/products/13231) and [that](https://www.sparkfun.com/products/13678)
Average delivery 2-4 weeks, may take up to 6 weeks.

So I should eventually figure it out.

**Comment by asetyde** [comment]
Hi , i've also a Nodemcu 1.0 , esp8266 12e family , with my test nothing .. i use this distribution that i think it's quite good .

many thanks for your library @bblanchon  , i tried to parse my struct to network but between c and java not so friendly game ... at now i start with your library . the best is take and parse a struct automatically in encode and also decode

**Comment by bblanchon** [comment]
@asetyde, thanks for your feedback. i'm glad to hear that it works on your side.

**Comment by asetyde** [comment]
sorry not , i do an error with example code , about this :  

StaticJsonBuffer<200> jsonBuffer;     200 what ? byte bit kbyte ?

**Comment by bblanchon** [comment]
These are bytes, [please have a look at the documentation](https://github.com/bblanchon/ArduinoJson/wiki/Memory%20model#introducing-staticjsonbuffer).

You can try to increase the size of the buffer, but 200 bytes should be enough for the example above.

**Comment by asetyde** [comment]
now i've added , they can , i hope , help us .. i think

**Comment by asetyde** [comment]
thanks ok  @bblanchon

**Comment by asetyde** [comment]
ok, wait Dynamic allocation is ok .

With static , how you determinate a static allocation in memory ... esp is not as arduino , has flash size distributed in this configuration :   

![screen shot 2015-08-31 at 10 47 24](https://cloud.githubusercontent.com/assets/6162152/9575172/af8c5b6e-4fcd-11e5-972f-34cf89d9db5e.png)

**Comment by bblanchon** [comment]
If dynamic allocation is OK to you, you can replace `StaticJsonBuffer` by `DynamicJsonBuffer`.

**Comment by asetyde** [comment]
yes , i've done . also when i tried different buffer size , there are strange but different result .

**Comment by bblanchon** [comment]
Can you be more specific?

**Comment by asetyde** [comment]
So so :)

i don't have many time to try , but i think the problem is where you allocate your static object , because there are many different about memory organization and space from esp and avr ... family is different .

when i 've tried @JeffAshtonCT  sketch with different size of buffer , reply on serial is many different  :  

also i think it's not good idea do this in setup function ...

i can do 3 fast example : 

with 200 you see stack error but , with 10 

![screen shot 2015-08-31 at 11 03 51](https://cloud.githubusercontent.com/assets/6162152/9575431/03f73712-4fd0-11e5-8c01-61a0dbe1c774.png)


with 50 

![screen shot 2015-08-31 at 11 04 55](https://cloud.githubusercontent.com/assets/6162152/9575449/37e74e9a-4fd0-11e5-9c88-e73fedbe5f10.png)

with 99 

![screen shot 2015-08-31 at 11 05 33](https://cloud.githubusercontent.com/assets/6162152/9575452/3ec28bc6-4fd0-11e5-8bb7-0b19fb2545bd.png)

**Comment by asetyde** [comment]
but is stupid test ..

**Comment by bblanchon** [comment]
Alexandro, it looks like you're using version 5.0.1. Can you checkout
master branch and try again?

Also, what happens with DynamicJsonBuffer?

I think that the stack may be too small. Do you know if you can change it?
This is usualy done in the compiler command line.
Le 31 août 2015 11:12, "Alexandro Todeschini" <notifications@github.com> a
écrit :

**Comment by igrr** [comment]
Hi all,
esp8266/Arduino guy here.

Exception is caused by an unaligned 32-bit store in `ArduinoJson::JsonBuffer::createObject`. The unaligned pointer comes from [placement new operator in JsonBufferAllocated](https://github.com/bblanchon/ArduinoJson/blob/master/include/ArduinoJson/Internals/JsonBufferAllocated.hpp#L16) which leads to `JsonBuffer::alloc`. For this particular sample posted by @JeffAshtonCT, this calls `StaticJsonBuffer::alloc`. Apparently `alloc` happily returns unaligned pointers which is okay on AVR but leads to a CPU exception on this particular core. Padding all allocations to 4 bytes inside `alloc` fixes this issue.

```c++
virtual void* alloc(size_t bytes) {
#ifdef ESP8266
bytes = (bytes + 3) & (~3);
#endif
...
```
Something similar is also required in `BlockJsonBuffer::allocInHead`, but i haven't investigated further.

**Comment by bblanchon** [comment]
Thanks Ivan. I'll fix that ASAP.
Le 31 août 2015 11:47, "Ivan Grokhotkov" <notifications@github.com> a
écrit :

**Comment by asetyde** [comment]
Russian is the best :)  :+1:  thanks @igrr 

@bblanchon  when you are ok , i'll try

**Comment by asetyde** [comment]
Dynamicbuffer is ok , work

**Comment by bblanchon** [comment]
I fixed the alignment in branch [alignment](https://github.com/bblanchon/ArduinoJson/tree/alignment), it should work on the ESP8266.
I'll try to reduce the code size before merging it into master.
Please give it a try.

**Comment by asetyde** [comment]
but you 've merged or not ?

**Comment by bblanchon** [comment]
No I didn't.
I'm not satisfied with the code size.
Did you try the "alignment" branch?

**Comment by asetyde** [comment]
also dynamic has stack exception with a heavy use to transfer a struct (2k bytes)

**Comment by asetyde** [comment]
wait , now i'll try

**Comment by asetyde** [comment]
Pass ok ; 

result : 
100
![screen shot 2015-09-01 at 18 27 42](https://cloud.githubusercontent.com/assets/6162152/9610011/2c61a9fa-50d7-11e5-940c-cf7c7c64beb6.png)
_______________
200 (as example)
![screen shot 2015-09-01 at 18 28 24](https://cloud.githubusercontent.com/assets/6162152/9610025/4076e23e-50d7-11e5-80f7-2589d115227b.png)
_____________
300
![screen shot 2015-09-01 at 18 29 29](https://cloud.githubusercontent.com/assets/6162152/9610058/69fc0c60-50d7-11e5-9ef8-b660f9ac84f5.png)

**Comment by bblanchon** [comment]
Thanks, I'll merge as soon as I find a way to get the code smaller.

**Comment by asetyde** [comment]
ok , a fast question , 
i create a son from reading by my struct , punt on char[] and after convert to char byte to send on wifiudp esp after i want to rebuild on java ..is possible more simply ?
and to know size of json

**Comment by bblanchon** [comment]
I saw issue #110, but I don't understand what you want to do.
Please keep discussion separated, and comment in #110.

**Comment by asetyde** [comment]
ok

**Comment by bblanchon** [comment]
It's fixed in master and will soon be released as 5.0.2.

**Comment by asetyde** [comment]
@igrr  Please , i not understand why memory space is too small , in BlockJsonBuffer , only parameter i see is :
static const size_t FIRST_BLOCK_CAPACITY = 32;

it can be has more memory  ?
or in dynamic use max esp capacity 

`#pragma once

#include "../JsonBuffer.hpp"

#include <stdlib.h>

namespace ArduinoJson {
namespace Internals {
class DefaultAllocator {
 public:
  void* allocate(size_t size) { return malloc(size); }
  void deallocate(void* pointer) { free(pointer); }
};

template <typename TAllocator>
class BlockJsonBuffer : public JsonBuffer {
  struct Block;
  struct EmptyBlock {
    Block* next;
    size_t capacity;
    size_t size;
  };
  struct Block : EmptyBlock {
    uint8_t data[1];
  };

 public:
  BlockJsonBuffer() : _head(NULL) {}

  ~BlockJsonBuffer() {
    Block* currentBlock = _head;

    while (currentBlock != NULL) {
      Block* nextBlock = currentBlock->next;
      _allocator.deallocate(currentBlock);
      currentBlock = nextBlock;
   `
